lib/rpc.ts

```diff
@@ -129,18 +131,31 @@ export class Rpc extends EventEmitter implements IForwarderDest {
   }
 
   /**
-   * Until start() is called, received messages are queued. This gives you an opportunity to
-   * register implementations and add "message" listeners without the risk of missing messages,
+   * Until start() is called, received and sent messages are queued. This gives you an opportunity
+   * to register implementations and add "message" listeners without the risk of missing messages,
    * even if receiveMessage() has already started being called.
    */
   public start(sendMessage: SendMessageCB) {
-    this._sendMessage = sendMessage;
+    this._sendMessageCB = sendMessage;
     if (this._inactiveQueue) {
       for (const msg of this._inactiveQueue) {
         this._dispatch(msg);    // We need to be careful not to throw from here.
       }
       this._inactiveQueue = null;
     }
+    for (const msg of this._inactiveSendQueue) {
+      sendMessage(msg);
```

Comment:
If sendMessage() throws ane exception (which is possible), it would leave us in an inconsistent state, with _inactiveSendQueue non-empty and containing messages that have already been sent. The receive-queue case suffers from a similar problem (see comment above), so we've been careful to handle exceptions when processing received messages. But those are potentially unavoidable when sending.

I propose for both queues the pattern:
```
   processQueue(this._inactiveSendQueue, sendMessage)
   function processQueue(queue, processFunc) {
      let i = 0;
      try {
        for ( ; i < queue.length; i++) {
          processFunc(queue[i]);
        }
      } finally {
         queue.splice(0, i);
      }
    }
```
Unfortunately this is sufficiently non-trivial to need a test case.
Also, if there is an exception, the best I can think of is that start() would propagate it and leave status unchanged (i.e. still not started). In other words, setting `this._sendMessage` would be at the end rather than at the start.

Reply:
I propose the following code which I think does the same thing and looks simpler too me: 
```
  public start(sendMessage: SendMessageCB) {
    this._sendMessageCB = sendMessage;
    try {
      // dispatch message received while inactive
      while (this._inactiveRecvQueue.length) {
        this._dispatch(this._inactiveRecvQueue.shift()!);
      }
      // send messages sent while inactive
      while (this._inactiveSendQueue.length) {
        sendMessage(this._inactiveSendQueue.shift()!);
      }
    } catch {
      // If an exception is thrown, lets propagate and leave state unchanged.
      this._sendMessageCB = null;
    }
  }
```
Note that `_sendMessageCB` is set first and then revert to null in case of failure and not only set at the end because `_dispatch` may need it to process some messages.